src/optimise/train.jl

```diff
@@ -80,6 +80,32 @@ end
 batchmemaybe(x) = tuple(x)
 batchmemaybe(x::Tuple) = x
 
+"""
+    step!(loss, params, opt)
+
+`step!` uses a `loss` function (with no inputs) to improve the [Model parameters](@ref) (`params`)
+based on a pluggable [Optimisers](@ref) (`opt`). It represents a single step in
+the training loop `train!`. While there is a default implementation for
+optimisers which are based on the `update!` function and only require gradient
+information, this `step!` has to be overloaded for more general optimisers.
+
+While the loss function of `train!` still accepts data as input, the loss function
+of `step!` accepts no input. `train!` cycles through the data in a loop
+roughly like this
+
+```julia
+for d in data
+  step!(ps, opt) do
+    loss(d)
+  end
+```
+
+"""
+function step!(loss, params, opt)
+  gs = gradient(loss, params)
+  update!(opt, params, gs)
+end
```

Comment:
I would prefer this to be consistent with `train!`, and so users can use the default step in a custom loop without knowing `batchmemaybe`.

```suggestion
function step!(loss, params, data, opt)
  gs = gradient(params) do
    loss(batchmemaybe(data)...)
  end
  update!(opt, params, gs)
end
```

Reply:
To be honest I do not know what `batchmaybe` does and took the view of someone wanting to write a new optimiser (i.e. overloading `step!`). In that case I would not want to deal with batchmaybe, so I left it in `train!`. Assuming that end user would use `train!` anyway. I guess the first question is, what does batchmaybe do?

Besides that, I wonder if it is cleaner to leave data out of the step!, this would essentially split the concern of optimisation (i.e. implementing a `step!`) from handling data, i.e. defining a loss. I wonder why `train!` doesn't have batch or shuffling utilities though, as that would be the data handling part. Without that data handling functionality you might as well just use a for loop

Reply:
`batchmemaybe` just turns things that are not a tuple into one. It just another way of doing:
```julia
if data isa Tuple
  loss(data...)
else
  loss(data)
end
```

It's worth noting that the main use-case (i.e. 99%) for `step!` is not overloading it, but writing a custom `train!`:
```julia
function mytrain!(loss, params, data, opt)
  # do some stuff
  for d in data
    # do other stuff
    # okay now update my model like Flux does
    # I don't want to worry about the details
    step!(loss, params, data, opt)
    # do some more stuff
  end
end
```
Seems weird if `Flux.train!` and `mytrain!` can't accept the same loss cause `Flux.train!`/`step!` don't treat the loss in a consistent way.

Reply:
The note about why not use a for-loop vs. make `train!` more complex is heavily discussed. The maintainers are split between "`train!` is a mistake" and "we should add more to `train!`." I'd recommend let's just avoid diving into that hole for this PR and assume that `train!` is behaviorally unchanged.

Reply:
Do you think that could be a callback for the `train!` function? "do other stuff" in every step?

Reply:
okay so `batchmaybe` does not help partitioning the data. And that seems like it really should be a functionality of `train!`. Doing stuff like

```
partition(1:size(x_train, ndims(x_train)), batchsize)
```

to get a generator over the index ranges of your batches is annoying. And this is just the simplest minibatching version without shuffling. At least something should deal with this data handling, and this is where this `batchmaybe` function really belongs. It should not be a concern of the optimiser

Reply:
Okay, thinking about it more, it is reasonable that the person writing a custom train loop should know what `d in data` looks like and how to pass it to the loss function. I'm convinced this is OK, let's just solicit some opinions from other maintainers since this is introducing a new API.

@mcabbott @DhairyaLGandhi @CarloLucibello @ToucheSir any thoughts?

Reply:
One other suggestion is to use `Zygote.withgradient` instead of `gradient` and return both the loss and gradients from `step!`.

Reply:
Bit of a drive-by but: Do we want to be adding new things which rely on zero-parameter functions and implicit Params? Aren't we trying to kill that entirely & move on?

Reply:
I thought about that, and right now, `train!` implicitly forces this. So, whenever we actually move to explicit, both `train!` and `step!` will be updated accordingly. Meaning this function would be `step!(loss, opt, args...)` and expect `loss(args...)`.

`train!` would be something like `loss(batchmemaybe(d)..., args...)` which is pretty hairy. That makes me think that the transition to explicit might be the right point to eliminate `train!` (fallout being that this PR would be irrelevant too).

Reply:
But also, it is pretty standard to write your loss function as `loss(x, y, m) = ...` so maybe not really that hairy other than the docstring.

Reply:
yeah, if the implicit params are gone, it would probably look like

```julia
function train!(...)
    for loss_sample in data_manager(loss, data)
        # data_manager generates from loss(params, x, y) a loss_sample(params) generator
        step!(loss, params, opt) # optimiser plugs params into loss
    end
end
```
So the same interface as now

Reply:
> One other suggestion is to use `Zygote.withgradient` instead of `gradient` and return both the loss and gradients from `step!`.

That is a good idea, what function should I use instead of `update!`then though?

Reply:
You would still use `update!`. `withgradient` returns two items: the loss and the gradients. The latter can be fed into `update!` as normal.

Reply:
so if we discard the loss anyway, why would you use withgradient?

Reply:
Sorry the suggestion is not to discard. We should return it along with the gradients at the end of `step!`. This will be useful for the custom loop case.